In [1]:
import pandas as pd
import wbgapi as wb

### Needed Series from WDI Database 
- EN.ATM.CO2E.KT - CO2 emissions (kt)
- EN.ATM.CO2E.PC - CO2 emissions (metric tons per capita)
- EN.ATM.CO2E.PP.GD.KD	- CO2 emissions (kg per 2017 PPP dollar of GDP)
- EG.USE.PCAP.KG.OE	 - Energy Use (kg of oil equivalent per capita)
- EG.FEC.RNEW.ZS - Renewable energy consumption (% of total final energy consumption)
- AG.YLD.CREL.KG - Cereal yield
- EG.ELC.ACCS.ZS - Access to Electricity (% of population)
- AG.LND.FRST.ZS - Forest area (% of land area)
- NY.GDP.MKTP.KD.ZG - GDP growth (annual %)
- NY.GDP.PCAP.CD - GDP per capita (current USD)
- NY.GNP.PCAP.CD - GNI per capita, Atlas method (current USD)
- SP.POP.TOTL - Total Population
- SP.POP.GROW - Population growth (annual %)
- SP.URB.GROW - Urban population growth (annual %)
- SP.URB.TOTL - Urban population
- SP.URB.TOTL.IN.ZS	- Urban population (% of total population)

## Extract emissions data from World Bank database via API (wbgapi)
### Ensure conda environment has wbgapi installed (pip install wbgapi)

In [2]:
series_list = ['EN.ATM.CO2E.KT',
               'EN.ATM.CO2E.PC',
               'EN.ATM.CO2E.PP.GD.KD',
               'EG.USE.PCAP.KG.OE',
               'EG.FEC.RNEW.ZS',
               'AG.YLD.CREL.KG',
               'EG.ELC.ACCS.ZS',
               'AG.LND.FRST.ZS',
               'NY.GDP.MKTP.KD.ZG',
               'NY.GDP.PCAP.CD',
               'NY.GNP.PCAP.CD',
               'SP.POP.TOTL',
               'SP.POP.GROW',
               'SP.URB.GROW',
               'SP.URB.TOTL',
               'SP.URB.TOTL.IN.ZS']

In [3]:
# Using wbgapi to extract World Bank data as Pandas data frame
raw_df = wb.data.DataFrame(series_list, time=range(1990, 2018), numericTimeKeys=True, labels=True, columns='series').reset_index()
raw_df.head()

economy  time   Country  Time  AG.LND.FRST.ZS  AG.YLD.CREL.KG  \
0     ZWE  2017  Zimbabwe  2017       45.451183          1202.7   
1     ZWE  2016  Zimbabwe  2016       45.570273           435.1   
2     ZWE  2015  Zimbabwe  2015       45.689363           557.5   
3     ZWE  2014  Zimbabwe  2014       45.808453           831.4   
4     ZWE  2013  Zimbabwe  2013       45.927543           668.5   

   EG.ELC.ACCS.ZS  EG.FEC.RNEW.ZS  EG.USE.PCAP.KG.OE  EN.ATM.CO2E.KT  \
0       44.178635           82.46                NaN    10340.000153   
1       42.561729           81.90                NaN    11020.000458   
2       33.700001           80.82                NaN    12430.000305   
3       32.299999           80.27                NaN    12079.999924   
4       40.498375           78.87         832.572236    12279.999733   

   EN.ATM.CO2E.PC  EN.ATM.CO2E.PP.GD.KD  NY.GDP.MKTP.KD.ZG  NY.GDP.PCAP.CD  \
0        0.700965              0.300613           4.080264     1192.107012   
1        0.762487              0.333455           0.900955     1421.787789   
2        0.878139              0.379509           2.023650     1410.329174   
3        0.871840              0.376287           1.484543     1407.034293   
4        0.905911              0.388196           3.196731     1408.367810   

   NY.GNP.PCAP.CD  SP.POP.GROW  SP.POP.TOTL  SP.URB.GROW  SP.URB.TOTL  \
0          1170.0     2.043620   14751101.0     1.860765    4755312.0   
1          1200.0     2.081806   14452704.0     1.806610    4667645.0   
2          1220.0     2.136294   14154937.0     1.769505    4584076.0   
3          1210.0     2.191391   13855753.0     1.730983    4503674.0   
4          1200.0     2.163267   13555422.0     1.613531    4426387.0   

   SP.URB.TOTL.IN.ZS  
0             32.237  
1             32.296  
2             32.385  
3             32.504  
4             32.654

In [4]:
# Rows and columns of data set
raw_df.shape

(7448, 20)

In [5]:
# # Datatypes of columns
raw_df.dtypes

economy                  object
time                      int64
Country                  object
Time                     object
AG.LND.FRST.ZS          float64
AG.YLD.CREL.KG          float64
EG.ELC.ACCS.ZS          float64
EG.FEC.RNEW.ZS          float64
EG.USE.PCAP.KG.OE       float64
EN.ATM.CO2E.KT          float64
EN.ATM.CO2E.PC          float64
EN.ATM.CO2E.PP.GD.KD    float64
NY.GDP.MKTP.KD.ZG       float64
NY.GDP.PCAP.CD          float64
NY.GNP.PCAP.CD          float64
SP.POP.GROW             float64
SP.POP.TOTL             float64
SP.URB.GROW             float64
SP.URB.TOTL             float64
SP.URB.TOTL.IN.ZS       float64
dtype: object

In [6]:
# Descriptive statistics
raw_df.describe()

time  AG.LND.FRST.ZS  AG.YLD.CREL.KG  EG.ELC.ACCS.ZS  \
count  7448.00000     7117.000000     6175.000000     6305.000000   
mean   2003.50000       32.625391     2894.700192       79.933424   
std       8.07829       23.388023     2336.051682       29.463801   
min    1990.00000        0.000000        0.100000        0.533899   
25%    1996.75000       12.513950     1404.050000       65.926689   
50%    2003.50000       30.855176     2388.700000       98.300003   
75%    2010.25000       47.617367     3796.250000      100.000000   
max    2017.00000       98.574551    36761.900000      100.000000   

       EG.FEC.RNEW.ZS  EG.USE.PCAP.KG.OE  EN.ATM.CO2E.KT  EN.ATM.CO2E.PC  \
count     7089.000000        4751.000000    6.669000e+03     6669.000000   
mean        31.159401        2270.160374    9.927095e+05        4.210340   
std         30.118135        2669.532102    3.204385e+06        5.229999   
min          0.000000           9.579196    0.000000e+00        0.000000   
25%          4.785963         603.049194    2.230000e+03        0.639336   
50%         20.946516        1238.114597    2.374000e+04        2.421594   
75%         54.760000        3025.736971    2.464900e+05        6.218103   
max         98.340000       21420.628504    3.351454e+07       47.651306   

       EN.ATM.CO2E.PP.GD.KD  NY.GDP.MKTP.KD.ZG  NY.GDP.PCAP.CD  \
count           6206.000000        6782.000000     6944.000000   
mean               0.268540           3.581076    11171.752248   
std                0.215092           5.787170    18975.239473   
min                0.000000         -64.047107       22.850371   
25%                0.138454           1.545476     1007.129242   
50%                0.215263           3.712218     3322.033110   
75%                0.330033           5.902814    13096.144542   
max                2.085052         149.972963   203266.913745   

       NY.GNP.PCAP.CD  SP.POP.GROW   SP.POP.TOTL  SP.URB.GROW   SP.URB.TOTL  \
count     6393.000000  7418.000000  7.420000e+03  7363.000000  7.364000e+03   
mean      9479.314154     1.478696  2.583084e+08     2.226490  1.210294e+08   
std      15075.519703     1.650633  8.196081e+08     2.092501  3.806065e+08   
min         40.000000   -27.722225  9.182000e+03   -27.707932  3.733000e+03   
25%        910.000000     0.546640  1.330466e+06     0.774823  6.515722e+05   
50%       3020.000000     1.408416  8.483160e+06     2.199913  4.041233e+06   
75%      10400.000000     2.416364  5.593234e+07     3.503700  3.182758e+07   
max     122130.000000    19.360429  7.578158e+09    31.143425  4.147419e+09   

       SP.URB.TOTL.IN.ZS  
count        7364.000000  
mean           55.266415  
std            23.529375  
min             5.416000  
25%            35.295750  
50%            54.165000  
75%            74.308513  
max           100.000000

## Data clean and prep

In [7]:
# Assign original dataframe to another that can be modified
emissions_df = raw_df
emissions_df.sample(5)

economy  time          Country  Time  AG.LND.FRST.ZS  AG.YLD.CREL.KG  \
3161     KOR  1992      Korea, Rep.  1992       67.758656     5937.600000   
4094     GAB  2011            Gabon  2011       91.735662     1596.100000   
784      CHE  2017      Switzerland  2017       31.855174     6800.200000   
4979     CHI  1994  Channel Islands  1994             NaN             NaN   
6293     SST  1996     Small states  1996       41.623449     1470.168398   

      EG.ELC.ACCS.ZS  EG.FEC.RNEW.ZS  EG.USE.PCAP.KG.OE  EN.ATM.CO2E.KT  \
3161       99.942215        0.752920        2534.341028   297050.000000   
4094       88.859566       79.710000        2308.796615     5780.000210   
784       100.000000       24.990000                NaN    38700.000763   
4979      100.000000             NaN                NaN             NaN   
6293             NaN       27.709437                NaN   114323.477719   

      EN.ATM.CO2E.PC  EN.ATM.CO2E.PP.GD.KD  NY.GDP.MKTP.KD.ZG  NY.GDP.PCAP.CD  \
3161        6.790031              0.465372           6.198643     8126.670390   
4094        3.260931              0.230525           7.091753    10273.799012   
784         4.578885              0.066261           1.362800    82254.376927   
4979             NaN                   NaN                NaN             NaN   
6293        4.054800                   NaN                NaN     3613.328087   

      NY.GNP.PCAP.CD  SP.POP.GROW  SP.POP.TOTL  SP.URB.GROW  SP.URB.TOTL  \
3161          8310.0     1.039161   43747962.0     2.163903   33169705.0   
4094          7720.0     3.525162    1772500.0     4.176634    1525981.0   
784          81870.0     0.933156    8451840.0     0.962986    6234162.0   
4979             NaN     0.122912     142466.0    -0.476620      43746.0   
6293             NaN     1.568983   28194606.0     2.384903   13377765.0   

      SP.URB.TOTL.IN.ZS  
3161          75.820000  
4094          86.092000  
784           73.761000  
4979          30.706000  
6293          47.447964

In [8]:
#Remove Time column as it is not neccessary column
emissions_df.drop(columns = ['Time'], axis = 1, inplace = True)

In [9]:
# Rename the columns headers to meaningful names
column_names = {'AG.LND.FRST.ZS':'forest_area_percent',
             'AG.YLD.CREL.KG':'cereal_yield',
             'EG.ELC.ACCS.ZS':'electricity_access_percent',
             'EG.FEC.RNEW.ZS':'renew_energy_percent',
             'EG.USE.PCAP.KG.OE':'energy_use_per_capita',
             'EN.ATM.CO2E.KT':'emissions_total',
             'EN.ATM.CO2E.PC':'emissions_per_capita',
             'EN.ATM.CO2E.PP.GD.KD':'emissions_per_gdp',
             'NY.GDP.MKTP.KD.ZG':'gdp_growth_percent',
             'NY.GDP.PCAP.CD':'gdp_per_capita',
             'NY.GNP.PCAP.CD':'gni_per_capita',
             'SP.POP.GROW':'pop_growth_percent',
             'SP.POP.TOTL':'pop_total',
             'SP.URB.GROW':'urb_pop_growth_percent',
             'SP.URB.TOTL':'urban_pop_total',
             'SP.URB.TOTL.IN.ZS':'urban_pop_percent',
             'economy':'country_code',
             'time':'year',
             'Country':'country_name'    
            }

In [10]:
emissions_df = emissions_df.rename(columns=column_names)
emissions_df.sample(10)

country_code  year              country_name  forest_area_percent  \
2713          MYS  1992                  Malaysia            62.191660   
2749          MDG  2012                Madagascar            21.546325   
1965          MNP  2012  Northern Mariana Islands            65.191304   
5865          ARG  2004                 Argentina            11.734029   
1497          WSM  2004                     Samoa            59.852297   
4187          FIN  2002                   Finland            73.557609   
1810          PNG  1999          Papua New Guinea            80.135445   
5451          BTN  1998                    Bhutan            64.978492   
2752          MDG  2009                Madagascar            21.681826   
3612          HUN  2017                   Hungary            22.542954   

      cereal_yield  electricity_access_percent  renew_energy_percent  \
2713        2953.1                         NaN             10.266217   
2749        3348.9                   18.700001             83.400000   
1965           NaN                  100.000000              0.000000   
5865        3658.8                   96.653503              9.290000   
1497           NaN                   90.728996             48.820000   
4187        3295.0                  100.000000             29.520000   
1810        3981.1                    6.426855             68.710840   
5451        1559.5                         NaN             91.531471   
2752        3183.4                   17.400000             87.070000   
3612        5835.5                  100.000000             14.540000   

      energy_use_per_capita  emissions_total  emissions_per_capita  \
2713            1603.110493     67170.000000              3.625577   
2749                    NaN      2740.000010              0.119306   
1965                    NaN              NaN                   NaN   
5865            1720.311256    141380.004883              3.656178   
1497             297.756722       189.999998              1.010246   
4187            6728.581405     63439.998627             12.198597   
1810                    NaN      2600.000000              0.488594   
5451                    NaN       200.000000              0.358974   
2752                    NaN      1690.000057              0.080030   
3612                    NaN     46439.998627              4.744602   

      emissions_per_gdp  gdp_growth_percent  gdp_per_capita  gni_per_capita  \
2713           0.303059            8.885118     3193.635381          2950.0   
2749           0.079696            3.011148      504.173730           480.0   
1965                NaN            0.810811    14247.789301             NaN   
5865           0.203685            9.029573     4258.160261          3360.0   
1497           0.201011            3.087260     2168.038807          1830.0   
4187           0.294375            1.707149    26997.752990         25620.0   
1810           0.151001            1.855554      653.408134           680.0   
5451           0.092990            5.914031      652.350989           530.0   
2752           0.051753           -3.978709      455.407381           460.0   
3612           0.160828            4.271976    14621.239596         13180.0   

      pop_growth_percent   pop_total  urb_pop_growth_percent  urban_pop_total  \
2713            2.787186  18526708.0                5.205499        9599428.0   
2749            2.728141  22966240.0                4.694565        7630993.0   
1965           -0.307021     52359.0               -0.219870          47703.0   
5865            1.015337  38668796.0                1.209159       34747780.0   
1497            0.337139    188073.0               -0.780000          40355.0   
4187            0.242381   5200598.0                0.406147        4290649.0   
1810            3.490894   5321388.0                2.198390         711736.0   
5451            1.861997    557143.0                6.126642         130243.0   
2752            2.899473  2

In [11]:
#swap year and country_name columns
col_list = list(emissions_df.columns)
x, y = col_list.index('year'), col_list.index('country_name')
col_list[y], col_list[x] = col_list[x], col_list[y]
emissions_df = emissions_df[col_list]
list(emissions_df.columns)

['country_code',
 'country_name',
 'year',
 'forest_area_percent',
 'cereal_yield',
 'electricity_access_percent',
 'renew_energy_percent',
 'energy_use_per_capita',
 'emissions_total',
 'emissions_per_capita',
 'emissions_per_gdp',
 'gdp_growth_percent',
 'gdp_per_capita',
 'gni_per_capita',
 'pop_growth_percent',
 'pop_total',
 'urb_pop_growth_percent',
 'urban_pop_total',
 'urban_pop_percent']

In [12]:
emissions_df.shape

(7448, 19)

In [13]:
emissions_df.dtypes

country_code                   object
country_name                   object
year                            int64
forest_area_percent           float64
cereal_yield                  float64
electricity_access_percent    float64
renew_energy_percent          float64
energy_use_per_capita         float64
emissions_total               float64
emissions_per_capita          float64
emissions_per_gdp             float64
gdp_growth_percent            float64
gdp_per_capita                float64
gni_per_capita                float64
pop_growth_percent            float64
pop_total                     float64
urb_pop_growth_percent        float64
urban_pop_total               float64
urban_pop_percent             float64
dtype: object

In [14]:
emissions_df.isnull().sum()

country_code                     0
country_name                     0
year                             0
forest_area_percent            331
cereal_yield                  1273
electricity_access_percent    1143
renew_energy_percent           359
energy_use_per_capita         2697
emissions_total                779
emissions_per_capita           779
emissions_per_gdp             1242
gdp_growth_percent             666
gdp_per_capita                 504
gni_per_capita                1055
pop_growth_percent              30
pop_total                       28
urb_pop_growth_percent          85
urban_pop_total                 84
urban_pop_percent               84
dtype: int64

Objective: Remove as many empty cells as possible, while preserving the highest possible amount of columns and rows. Check missing data for certain years, countries or features (columns).

In [15]:
# Assigned to a new df for dealing with NaN
clean_df = emissions_df
clean_df.head()

country_code country_name  year  forest_area_percent  cereal_yield  \
0          ZWE     Zimbabwe  2017            45.451183        1202.7   
1          ZWE     Zimbabwe  2016            45.570273         435.1   
2          ZWE     Zimbabwe  2015            45.689363         557.5   
3          ZWE     Zimbabwe  2014            45.808453         831.4   
4          ZWE     Zimbabwe  2013            45.927543         668.5   

   electricity_access_percent  renew_energy_percent  energy_use_per_capita  \
0                   44.178635                 82.46                    NaN   
1                   42.561729                 81.90                    NaN   
2                   33.700001                 80.82                    NaN   
3                   32.299999                 80.27                    NaN   
4                   40.498375                 78.87             832.572236   

   emissions_total  emissions_per_capita  emissions_per_gdp  \
0     10340.000153              0.700965           0.300613   
1     11020.000458              0.762487           0.333455   
2     12430.000305              0.878139           0.379509   
3     12079.999924              0.871840           0.376287   
4     12279.999733              0.905911           0.388196   

   gdp_growth_percent  gdp_per_capita  gni_per_capita  pop_growth_percent  \
0            4.080264     1192.107012          1170.0            2.043620   
1            0.900955     1421.787789          1200.0            2.081806   
2            2.023650     1410.329174          1220.0            2.136294   
3            1.484543     1407.034293          1210.0            2.191391   
4            3.196731     1408.367810          1200.0            2.163267   

    pop_total  urb_pop_growth_percent  urban_pop_total  urban_pop_percent  
0  14751101.0                1.860765        4755312.0             32.237  
1  14452704.0                1.806610        4667645.0             32.296  
2  14154937.0                1.769505        4584076.0             32.385  
3  13855753.0                1.730983        4503674.0             32.504  
4  13555422.0                1.613531        4426387.0             32.654

### Filter years by missing values

In [16]:
#define an array with the unique year values
years_count_missing = dict.fromkeys(clean_df['year'].unique(), 0)
for ind, row in clean_df.iterrows():
    years_count_missing[row['year']] += row.isnull().sum()

# sort the years by missing values
years_missing_sorted = dict(sorted(years_count_missing.items(), key=lambda item: item[1]))

# print the missing values for each year
print("missing values by year:")
for key, val in years_missing_sorted.items():
    print(key, ":", val)


missing values by year:
2007 : 262
2006 : 264
2005 : 271
2013 : 274
2004 : 274
2014 : 277
2012 : 279
2011 : 283
2010 : 289
2009 : 293
2008 : 297
2003 : 308
2002 : 314
2001 : 334
2000 : 342
2015 : 417
1999 : 430
1998 : 442
1997 : 454
1996 : 474
1995 : 499
2017 : 508
2016 : 508
1994 : 539
1993 : 553
1992 : 587
1991 : 680
1990 : 687


Consider years 1995 - 2015  years with less than 510 missing year data.

In [17]:
print("number of missing values in the whole dataset before filtering the years:")
print(clean_df.isnull().sum().sum())
print("number of rows before filtering the years:")
print(clean_df.shape[0])

# filter only rows for years between 1991 and 2008 (having less missing values)
clean_df = clean_df[(clean_df['year'] >= 1995) & (clean_df['year'] <= 2015)]

print("number of missing values in the whole dataset after filtering the years:")
print(clean_df.isnull().sum().sum())
print("number of rows after filtering the years:")
print(clean_df.shape[0])

number of missing values in the whole dataset before filtering the years:
11139
number of rows before filtering the years:
7448
number of missing values in the whole dataset after filtering the years:
7077
number of rows after filtering the years:
5586


### Filter countries by missing value

In [18]:
# check the amount of missing values by country

# define an array with the unique country values
countries_count_missing = dict.fromkeys(clean_df['country_code'].unique(), 0)

# iterate through all rows and count the amount of NaN values for each country
for ind, row in clean_df.iterrows():
    countries_count_missing[row['country_code']] += row.isnull().sum()

# sort the countries by missing values
countries_missing_sorted = dict(sorted(countries_count_missing.items(), key=lambda item: item[1]))

# print the missing values for each country
print("missing values by country:")
for key, val in countries_missing_sorted.items():
    print(key, ":", val)

missing values by country:
USA : 0
GBR : 0
CHE : 0
SWE : 0
ESP : 0
SVK : 0
PRT : 0
POL : 0
NOR : 0
NZL : 0
NLD : 0
MEX : 0
KOR : 0
JPN : 0
ITA : 0
IRL : 0
HUN : 0
DEU : 0
FRA : 0
FIN : 0
DNK : 0
CZE : 0
CHL : 0
AUT : 0
AUS : 0
PST : 0
OED : 0
NAC : 0
HIC : 0
EUU : 0
EMU : 0
URY : 1
UKR : 1
TUN : 1
TZA : 1
SEN : 1
SAU : 1
RUS : 1
ROU : 1
PHL : 1
PER : 1
PRY : 1
PAN : 1
OMN : 1
MKD : 1
NGA : 1
NIC : 1
NAM : 1
MAR : 1
MUS : 1
MLT : 1
KWT : 1
KEN : 1
KAZ : 1
JOR : 1
JAM : 1
IDN : 1
IND : 1
HND : 1
HTI : 1
GTM : 1
GHA : 1
SLV : 1
EGY : 1
ECU : 1
DOM : 1
CYP : 1
CIV : 1
COL : 1
CMR : 1
BGR : 1
BRN : 1
BRA : 1
BWA : 1
BOL : 1
BLR : 1
BGD : 1
ARG : 1
ALB : 1
TSA : 1
SAS : 1
LMC : 1
TLA : 1
LAC : 1
LCN : 1
TEC : 1
ECA : 1
ECS : 1
EAR : 1
CEB : 1
AFW : 1
ZWE : 2
ZMB : 2
ZAF : 2
NPL : 2
BIH : 2
BEN : 2
TSS : 2
SSA : 2
SSF : 2
ARB : 2
UZB : 3
SVN : 3
MOZ : 3
KGZ : 3
ISR : 3
PRE : 3
VNM : 4
TGO : 4
PAK : 4
MDA : 4
LTU : 4
LVA : 4
HRV : 4
KHM : 4
WLD : 4
IDA : 4
IDX : 4
IDB : 4
TUR : 5
TJK : 5
BEL :

Remove countries with more than 100 missing values 

In [19]:
print("number of missing values in the whole dataset before filtering the countries:")
print(clean_df.isnull().sum().sum())
print("number of rows before filtering the countries:")
print(clean_df.shape[0])


# filter only rows for countries with less than 90 missing values
countries_filter = []
for key, val in countries_missing_sorted.items():
    if val<100:
        countries_filter.append(key)

clean_df = clean_df[clean_df['country_code'].isin(countries_filter)]

print("number of missing values in the whole dataset after filtering the countries:")
print(clean_df.isnull().sum().sum())
print("number of rows after filtering the countries:")
print(clean_df.shape[0])

number of missing values in the whole dataset before filtering the countries:
7077
number of rows before filtering the countries:
5586
number of missing values in the whole dataset after filtering the countries:
2941
number of rows after filtering the countries:
5019


### Filter years by missing values

In [20]:
clean_df.isnull().sum()

country_code                     0
country_name                     0
year                             0
forest_area_percent             27
cereal_yield                   384
electricity_access_percent     407
renew_energy_percent            30
energy_use_per_capita         1264
emissions_total                 51
emissions_per_capita            51
emissions_per_gdp              289
gdp_growth_percent             133
gdp_per_capita                  74
gni_per_capita                 231
pop_growth_percent               0
pop_total                        0
urb_pop_growth_percent           0
urban_pop_total                  0
urban_pop_percent                0
dtype: int64

Top 4 features with missing values (energy_use_per_capita, electricity_access_percent, cereal_yield, emissions_per_gdp)

In [21]:
# remove features with more than 25 missing values

from itertools import compress

# create a boolean mapping of features with more than 25 missing values
vars_bad = clean_df.isnull().sum()>25

# remove the columns corresponding to the mapping of the features with many missing values
clean_df2 = clean_df.drop(compress(data = clean_df.columns, selectors = vars_bad), axis='columns')

print("Remaining missing values per column:")
print(clean_df2.isnull().sum())

Remaining missing values per column:
country_code              0
country_name              0
year                      0
pop_growth_percent        0
pop_total                 0
urb_pop_growth_percent    0
urban_pop_total           0
urban_pop_percent         0
dtype: int64


In [22]:
clean_df2.shape

(5019, 8)

In [23]:
clean_df2.to_csv('../Resources/emissions_2.csv',index= False)

### Compare to DropNa on emissions_df

In [24]:
#Drop rows which has null values
emissions_df = emissions_df.dropna()
emissions_df.reset_index(drop = True)

country_code                 country_name  year  forest_area_percent  \
0             ZWE                     Zimbabwe  2013            45.927543   
1             ZWE                     Zimbabwe  2012            46.046633   
2             ZWE                     Zimbabwe  2011            46.165723   
3             ZWE                     Zimbabwe  2010            46.284813   
4             ZWE                     Zimbabwe  2009            46.403903   
...           ...                          ...   ...                  ...   
3547          AFE  Africa Eastern and Southern  2004            33.654274   
3548          AFE  Africa Eastern and Southern  2003            34.177441   
3549          AFE  Africa Eastern and Southern  2002            34.357452   
3550          AFE  Africa Eastern and Southern  2001            34.537463   
3551          AFE  Africa Eastern and Southern  2000            34.717473   

      cereal_yield  electricity_access_percent  renew_energy_percent  \
0       668.500000                   40.498375             78.870000   
1       695.700000                   44.000000             77.500000   
2       587.400000                   36.900002             79.270000   
3       733.400000                   38.782551             82.270000   
4       452.400000                   43.369083             82.090000   
...            ...                         ...                   ...   
3547   1394.049728                   23.815880             63.279292   
3548   1296.064498                   22.548307             64.165735   
3549   1453.381135                   21.601503             64.622779   
3550   1442.820237                   19.986220             65.586773   
3551   1459.724505                   19.957930             65.128796   

      energy_use_per_capita  emissions_total  emissions_per_capita  \
0                832.572236     12279.999733              0.905911   
1                814.910235     12010.000229              0.905368   
2                787.030033     11409.999847              0.875955   
3                736.691254      9600.000381              0.747677   
4                720.587138      7750.000000              0.611208   
...                     ...              ...                   ...   
3547             751.434195    457599.991269              1.027664   
3548             733.782234    426379.996829              0.982878   
3549             720.036948    404210.007127              0.956164   
3550             732.364434    393150.000000              0.954243   
3551             727.105591    355920.000000              0.886254   

      emissions_per_gdp  gdp_growth_percent  gdp_per_capita  gni_per_capita  \
0              0.388196            3.196731     1408.367810     1200.000000   
1              0.391797           15.744877     1290.193956     1120.000000   
2              0.430830           14.620207     1082.615774      950.000000   
3              0.415482           21.452061      937.840338      650.000000   
4              0.407369           12.019560      762.297957      440.000000   
...                 ...                 ...             ...             ...   
3547           0.334259            5.507315      985.530235      810.760703   
3548           0.329285            3.096045      812.946404      666.312226   
3549           0.322486            3.905250      626.559857      615.771592   
3550           0.326004            3.653224      628.204191      629.894844   
3551           0.306009            3.353978      706.992061      656.176751   

      pop_growth_percent    pop_total  urb_pop_growth_percent  \
0               2.163267   13555422.0                1.613531   
1               1.822309   13265331.0                1.272568   
2               1.438339   13025785.0                0.891612   
3               1.253650   12839771.0                0.706879   
4               1.026265   12679810.0                0.482488   
...                  ...          .

In [25]:
emissions_df.to_csv('../Resources/emissions.csv',index= False)

### Shape of data frame:
- Filter and Bin by missing values: (5019, 8)
- Drop (NA: 3552, 19)